python_modules/dagster/dagster/_core/definitions/asset_check_execution.py

```diff
@@ -0,0 +1,113 @@
+import enum
+from typing import TYPE_CHECKING, NamedTuple, Optional
+
+import dagster._check as check
+
+from .asset_check_result import AssetCheckResult
+from .events import AssetKey
+
+if TYPE_CHECKING:
+    from dagster._core.event_api import EventLogRecord
+
+
+class AssetCheckExecutionStatus(enum.Enum):
+    PLANNED = "PLANNED"  # No result yet
+    RESULT_SUCCESS = "RESULT_SUCCESS"
+    RESULT_FAILURE = "RESULT_FAILURE"
+
+
+class AssetCheckExecution(
+    NamedTuple(
+        "_AssetCheckExecution",
+        [
+            ("asset_key", AssetKey),
+            ("check_name", str),
+            ("check_status", AssetCheckExecutionStatus),
+            ("run_id", str),
+            ("check_result", Optional[AssetCheckResult]),
+            ("target_materialization_record", Optional["EventLogRecord"]),
+            ("start_timestamp", Optional[float]),
+            ("end_timestamp", Optional[float]),
+        ],
+    )
+):
+    """Wrapper around AssetCheckResult that includes metadata about the execution of the check.
+
+    Normal use is to instantiate this class with AssetCheckExecution.planned(), then call
+    with_result() when the check is complete.
+    """
+
+    def __new__(
+        cls,
+        asset_key: AssetKey,
+        check_name: str,
+        check_status: AssetCheckExecutionStatus,
+        run_id: str,
+        check_result: Optional[AssetCheckResult] = None,
+        target_materialization_record: Optional["EventLogRecord"] = None,
+        start_timestamp: Optional[float] = None,
+        end_timestamp: Optional[float] = None,
+    ):
+        if check_status == AssetCheckExecutionStatus.PLANNED:
+            check.invariant(
+                check_result is None, "check_result must be None when check_status is PLANNED"
+            )
+        else:
+            check_result = check.not_none(
```

Comment:
slightly annoying that we have to check all of this. required if we're going to create AssetCheckExecutions before the results come in.

Reply:
A slightly different option would be to have 3 objects:

1. the one returned from the user code. AssetCheckResult
2. one that wraps 1 with more metadata. AssetCheckResultData?
3. one that we write before the results come in, and optionally contains 2.

A clarifying thing of this approach is that the event specific data could use 2 instead of 3, since we don't write the event until the result happens. A negative is that I think we'd still have to expose 3 to users, since they may want to query about in progress checks. And they'd get an extra wrapper class in the mix